### PR TITLE
Fix some webui that doesn't allowed in private window is loaded in private window

### PR DESCRIPTION
**Pull request by simonhong**
Some webui should not be loaded in private window.
Instead, they should be redirected to normal window.
This also fixes crash when user tries to load sync page in private window.
Sync page will be also redirected to normal window.

Fix https://github.com/brave/brave-browser/issues/2853
Fix https://github.com/brave/brave-browser/issues/2852

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Windows
  - [x] macOS
  - [x] Linux
- Verified that these changes pass automated tests (`npm test brave_unit_tests && npm test brave_browser_tests`) on
  - [ ] Windows
  - [x] macOS
  - [x] Linux
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone

## Test Plan:
`BraveSchemeLoadBrowserTest.*PageIsNotAllowedInPrivateWindow`

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

### CODE REVIEW

File: patches/content-public-common-url_utils.cc.patch [hunk 1]
```diff
@@ -0,0 +1,33 @@
+diff --git a/content/public/common/url_utils.cc b/content/public/common/url_utils.cc
+index 5fcc9b77371c5f51a7911cfaaa5ba77287832223..040907551cf508ca642edf0d057d231745cd6d6d 100644
+--- a/content/public/common/url_utils.cc
++++ b/content/public/common/url_utils.cc
+@@ -20,6 +20,12 @@
+ namespace content {
+ 
+ bool HasWebUIScheme(const GURL& url) {
++#if defined(BRAVE_CHROMIUM_BUILD)
```
Comment (simonhong) [inline patches/content-public-common-url_utils.cc.patch]: Will try to use override strategy.

File: chromium_src/chrome/browser/ui/browser_navigator.cc [hunk 2]
```diff
@@ -5,5 +5,30 @@
...
 
 #define ChromeNavigationUIData BraveNavigationUIData
+#define Navigate Navigate_ChromiumImpl
+#define IsURLAllowedInIncognito IsURLAllowedInIncognito_ChromiumImpl
 #include "../../../../chrome/browser/ui/browser_navigator.cc"
 #undef ChromeNavigationUIData
+#undef Navigate
+#undef IsURLAllowedInIncognito
+
+void Navigate(NavigateParams* params) {
```
Comment (bridiver) [inline chromium_src/chrome/browser/ui/browser_navigator.cc]: this is exactly what we don't want to do: try to find every place we need to hook in to map the url. We need to map it once and only once in a place that takes effect before any security checks

Comment (simonhong) [reply chromium_src/chrome/browser/ui/browser_navigator.cc]: I tried to test this place as a replacement of current mapping not for another hooking with current one. So I inserted `NOTREACHED()` there.

Comment (simonhong) [reply chromium_src/chrome/browser/ui/browser_navigator.cc]: Of course, I'll still studying navigation flow to find better place :)

Comment (bridiver) [reply chromium_src/chrome/browser/ui/browser_navigator.cc]: yep, just need to remember that there are many, many different paths and most of the are initiated by the renderer

**Comment by simonhong** [comment]
@bridiver With this new mapping timing, previously skipped url check/verification that did for chrome ui are also applied to brave ui for browser-initiated navigation.
After https://github.com/brave/brave-core/pull/1196, renderer-initiated navigation doesn't use brave url. I spend a lot of time to investigate about navigation path and didn't find better place than this.
PTAL again.

File: browser/brave_content_browser_client.cc [hunk 3]
```diff
@@ -63,6 +63,12 @@ namespace {
 
 bool HandleURLRewrite(GURL* url,
                       content::BrowserContext* browser_context) {
+  if (url->SchemeIs(kBraveUIScheme)) {
```
Comment (bridiver) [inline browser/brave_content_browser_client.cc]: seems like this should be a DCHECK

File: patches/chrome-browser-ui-browser_navigator.cc.patch [hunk 4]
```diff
@@ -0,0 +1,21 @@
+diff --git a/chrome/browser/ui/browser_navigator.cc b/chrome/browser/ui/browser_navigator.cc
+index 57cc96c276c1106079828b5e359a3d88e1c901ce..81baaee0fdad10d4b1f4dc17889a44550083ee46 100644
+--- a/chrome/browser/ui/browser_navigator.cc
++++ b/chrome/browser/ui/browser_navigator.cc
+@@ -707,6 +707,16 @@ bool IsHostAllowedInIncognito(const GURL& url) {
+            host != chrome::kChromeUISuggestionsHost;
+   }
+ 
++#if defined(BRAVE_CHROMIUM_BUILD)
++  DCHECK(!url.SchemeIs(content::kBraveUIScheme))
```
Comment (bridiver) [inline patches/chrome-browser-ui-browser_navigator.cc.patch]: we definitely do not want to patch in a DCHECK here

File: chromium_src/chrome/browser/browser_about_handler.cc [hunk 5]
```diff
@@ -13,6 +13,10 @@
 bool FixupBrowserAboutURLBraveImpl(GURL* url,
                                    content::BrowserContext* browser_context) {
   if (url->SchemeIs(kBraveUIScheme)) {
+    NOTREACHED() << "brave url should not be reached here. scheme mapping "
```
Comment (bridiver) [inline chromium_src/chrome/browser/browser_about_handler.cc]: why would we put a NOTREACHED() before a bunch of code? This does not seem correct to me

File: patches/chrome-browser-ui-browser_navigator.cc.patch [hunk 6]
```diff
@@ -0,0 +1,16 @@
+diff --git a/chrome/browser/ui/browser_navigator.cc b/chrome/browser/ui/browser_navigator.cc
+index 57cc96c276c1106079828b5e359a3d88e1c901ce..aa184ea5a2881ebbce3a106e55ac864212329ce6 100644
+--- a/chrome/browser/ui/browser_navigator.cc
++++ b/chrome/browser/ui/browser_navigator.cc
+@@ -707,6 +707,11 @@ bool IsHostAllowedInIncognito(const GURL& url) {
+            host != chrome::kChromeUISuggestionsHost;
+   }
+ 
++#if defined(BRAVE_CHROMIUM_BUILD)
```
Comment (bridiver) [inline patches/chrome-browser-ui-browser_navigator.cc.patch]: why is this needed?

Comment (simonhong) [reply patches/chrome-browser-ui-browser_navigator.cc.patch]: This is needed for our sync page.
Our sync page is not allowed to load private window.
That causes crash for now.

Comment (bridiver) [reply patches/chrome-browser-ui-browser_navigator.cc.patch]: so this is a good example of what I brought up in slack today. If we create this one-off patch then someone else will have to come in and create another one if there is some other host we don't want in private tabs. We should have a chromium_src override that defines a method to call here

**Comment by simonhong** [comment]
@bridiver This PR will work well after merging https://github.com/brave/brave-core/pull/1573.
